### DISCUSSION TITLE
Using uart on pico

### DISCUSSION THREAD

**Discussion by didyblunder** [micropython/micropython#16419] [RP2040 / Pico] [answered]
I've not actually used the pico uart before (or uart on anything so forgive a noob question), but I have 2 pico's and wired their serial gpio pins together (rx/tx swapped and grounds connected)
The sender pico has the following:
```
from machine import UART, Pin
import time
uart0 = UART(0, baudrate=9600, tx=Pin(0), rx=Pin(1))
while True:
    uart0.write('hello')
    uart0.write('world\n')
    print('sent!')
    time.sleep(5)
```

if the receiver pico has the following (uart0.read() )
```
from machine import UART, Pin
import time
uart0 = UART(0, baudrate=9600, tx=Pin(0), rx=Pin(1))
while True:   
    if uart0.any():
        print(uart0.read())
```
I get:
MPY: soft reboot
b'helloworld\n'
b'helloworld\n'

If the receiver pico has the following (uart0.readline() )
```
from machine import UART, Pin
import time
uart0 = UART(0, baudrate=9600, tx=Pin(0), rx=Pin(1))
while True:   
    if uart0.any():
        print(uart0.readline())
```
I get:
MPY: soft reboot
b'h'
b'e'
b'l'
b'l'
b'o'
b'w'
b'o'
b'r'
b'l'
b'd'
b'\n'

I was expecting the printouts to be the other way round - i.e. the uart0.readline() would only print out once a '\n' was received, and the read would print out each character, but to my surprise its the other way round.

This is probably my misunderstanding, but I would appreciate it is someone would confirm the pico is working as expected, and thats the way the uart is supposed to work.  (or perhaps to point out that my test received code is flawed, as I cant find appropriate examples with a quick google of micropython uart)

**Comment by GitHubsSilverBullet** [comment, marked as answer]
You want to add:
timeout=: specifies the time to wait for the first character (in ms).
timeout_char=: specifies the time to wait between characters (in ms).
on the receiver side. Maybe 10ms…
Otherwise the UART doesn't wait for the rest of the line.
`uart0 = UART(0, baudrate=9600, tx=Pin(0), rx=Pin(1), timeout=10, timeout_char=10)`

**Comment by didyblunder** [comment]
Thanks for that quick response.  Putting in the timeouts you suggest makes the readline() work as I though it should.
